src/Dscp/Network/Wrapped.hs

```diff
@@ -0,0 +1,283 @@
+    (cEnv :: ClientEnv t) <- registerClient @t wId wMsgTypes wSubs
+    nat (wAction cEnv) `finally` (logDebug $ fromString $ "Worker " <> show wId <> " has exited")
+
+cliSend ::
+       forall t d m. (Message MsgK d, NetworkingCli t m, MonadIO m)
+    => ClientEnv t
+    -> Maybe (NodeId t)
+    -> d
+    -> m ()
+cliSend btq nId msg =
+    atomically $ sendBtq btq (nId, (msgType @d, [BSL.toStrict $ serialise msg]))
+
+data CliRecvExc
+    = CRETimeout
+    | CREUnexpected Text
+    | CREMalformed Text
+    deriving (Eq, Show, Generic)
+
+instance Exception CliRecvExc
+
+data CliRecvExcInternal
+    = CREMalformedTag
+    | CRENoCallback Natural
+    | CREWrongFramesNumber Int
+    deriving (Eq,Show,Generic)
+
+instance Exception CliRecvExcInternal
+
+type SendConstraint k t d m
+     = ( Message k d
+       , NetworkingCli t m
+       , MonadUnliftIO m
+       , MonadCatch m
+       , MonadLogging m)
+
+-- Timeout -- milliseconds, 0 if instant response is expected, -1 (any
+-- negative) if timeout should be disabled.
+--
+-- Callback takes care of decoding itself because we might want to
+-- propagate data first (before decoding the message).
+cliRecv ::
+       forall t k d m a. SendConstraint k t d m
+    => ClientEnv t
+    -> Int
+    -> [CallbackWrapper (NodeId t) m a]
+    -> m a
+cliRecv btq timeout callbacks = withHandler $ withTimeout $ \tmAction -> do
+    res <- atomically $ (Right <$> recvBtq btq) `orElse` (Left <$> tmAction)
+    let call nId msgTagM msgs = do
+            msgTag <- maybe (throwM CREMalformedTag) pure msgTagM
+            msg <- case msgs of
+                [x]   -> pure x
+                other -> throwM $ CREWrongFramesNumber (length other)
+            runCallbacksInt callbacks msgTag msg nId >>= \case
+                Nothing -> throwM $ CRENoCallback msgTag
+                Just x -> pure x
+    case res of
+        Right (nId, L.Response msgT msg) -> call nId (fromMsgType msgT) msg
+        Right (nId, L.Update sub msg)    -> call nId (fromSubType sub) msg
+        Left ()                          -> throwM CRETimeout
+  where
+    withHandler x =
+        catch x $ \(e :: CliRecvExcInternal) -> do
+            logWarning $ fromString $ "Could not receive: " <> show e <> ", retrying"
+            case e of
+                CRENoCallback n -> throwM $ CREUnexpected $ "No callback for " <> show n
+                -- we ignore messages related to malformed input from
+                -- the network (just warn)
+                _               -> withHandler x
+
+    withTimeout :: (STM () -> m x) -> m x
+    withTimeout action
+        | timeout < 0 = action retry
+        | otherwise = do
+              timeoutVar <- liftIO newEmptyTMVarIO
+              let timer = do
+                      liftIO $ threadDelay timeout
+                      atomically $ putTMVar timeoutVar ()
+              withAsync timer $ \_async -> action (readTMVar timeoutVar)
```

Comment:
How does it work? IIUC If call of `recvBtq` above sees empty queue, `timeoutVar` is performed and you block until `timeout` passes and try to throw `CRETimeout`, and immediatelly after that kill the action without waiting for its complection, why so?

Reply:
I doubt it works like this. In my view, STM's `orElse` takes two "blocking" actions -- `recvBtq` and `readTMVar` -- and then will give me the result which comes earlier. It'd be the other way if one of these actions was not blocking, of course.

I'll test this (should have done it anyway).

I'll test it anyway (I should have).